### PR TITLE
feat(coordinator): add a prover block list

### PR DESCRIPTION
**Pull request by colinlyguo**
### Purpose or design rationale of this PR

Add a prover block list.

Also, add a unit test (adding/removing a blocked public key) to test the effectiveness of the block list.

### PR title

Your PR title must follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) (as we are doing squash merge for each PR), so it must start with one of the following [types](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type):

- [x] feat: A new feature

### Deployment tag versioning

Has `tag` in `common/version.go` been updated or have you added `bump-version` label to this PR?

- [x] Yes

### Breaking change label

Does this PR have the `breaking-change` label?

- [x] No, this PR is not a breaking change

### CODE REVIEW

**Comment by 0xmountaintop** [comment]
can we make use of `CheckScrollProverVersion`?

**Comment by colinlyguo** [comment]
> CheckScrollProverVersion

Do you mean moving this check outside `if getTaskParameter.VK != b.vk {`?
```
	// if the prover has a different vk
	if getTaskParameter.VK != b.vk {
		// if the prover reports a different prover version
		if !version.CheckScrollProverVersion(proverVersion.(string)) {
			return nil, fmt.Errorf("incompatible prover version. please upgrade your prover, expect version: %s, actual version: %s", version.Version, proverVersion.(string))
		}
		// if the prover reports a same prover version
		return nil, fmt.Errorf("incompatible vk. please check your params files or config files")
	}
```
Yeah, we can do that. And also, I remembered some similar discussions before during `sepolia testnet` testing, this will make only the latest version available. also cc @georgehao @Thegaram

**Comment by 0xmountaintop** [comment]
> > CheckScrollProverVersion
> 
> Do you mean moving this check outside `if getTaskParameter.VK != b.vk {`?
> 
> ```
> 	// if the prover has a different vk
> 	if getTaskParameter.VK != b.vk {
> 		// if the prover reports a different prover version
> 		if !version.CheckScrollProverVersion(proverVersion.(string)) {
> 			return nil, fmt.Errorf("incompatible prover version. please upgrade your prover, expect version: %s, actual version: %s", version.Version, proverVersion.(string))
> 		}
> 		// if the prover reports a same prover version
> 		return nil, fmt.Errorf("incompatible vk. please check your params files or config files")
> 	}
> ```
> 
> Yeah, we can do that. And also, I remembered some similar discussions before during `sepolia testnet` testing, this will make only the latest version available. also cc @georgehao @Thegaram

I mean move this out and modify. so that we check version > XXX (this can solve "only the latest version available").

**Comment by colinlyguo** [comment]
> > > CheckScrollProverVersion
> > 
> > 
> > Do you mean moving this check outside `if getTaskParameter.VK != b.vk {`?
> > ```
> > 	// if the prover has a different vk
> > 	if getTaskParameter.VK != b.vk {
> > 		// if the prover reports a different prover version
> > 		if !version.CheckScrollProverVersion(proverVersion.(string)) {
> > 			return nil, fmt.Errorf("incompatible prover version. please upgrade your prover, expect version: %s, actual version: %s", version.Version, proverVersion.(string))
> > 		}
> > 		// if the prover reports a same prover version
> > 		return nil, fmt.Errorf("incompatible vk. please check your params files or config files")
> > 	}
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Yeah, we can do that. And also, I remembered some similar discussions before during `sepolia testnet` testing, this will make only the latest version available. also cc @georgehao @Thegaram
> 
> I mean move this out and modify. so that we check version > XXX (this can solve "only the latest version available").

oh, I see. then we can set the version as a config?

**Comment by colinlyguo** [comment]
> CheckScrollProverVersion

will change to this implementation.

**Comment by colinlyguo** [comment]
> > > CheckScrollProverVersion
> > 
> > 
> > Do you mean moving this check outside `if getTaskParameter.VK != b.vk {`?
> > ```
> > 	// if the prover has a different vk
> > 	if getTaskParameter.VK != b.vk {
> > 		// if the prover reports a different prover version
> > 		if !version.CheckScrollProverVersion(proverVersion.(string)) {
> > 			return nil, fmt.Errorf("incompatible prover version. please upgrade your prover, expect version: %s, actual version: %s", version.Version, proverVersion.(string))
> > 		}
> > 		// if the prover reports a same prover version
> > 		return nil, fmt.Errorf("incompatible vk. please check your params files or config files")
> > 	}
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Yeah, we can do that. And also, I remembered some similar discussions before during `sepolia testnet` testing, this will make only the latest version available. also cc @georgehao @Thegaram
> 
> I mean move this out and modify. so that we check version > XXX (this can solve "only the latest version available").

A question when implementing this method, the version compared in `CheckScrollProverVersion` is the prover version, does it have the feature of `monotonically increasing` based on some comparison rules?

code (the former part):
```
// ZkVersion is commit-id of common/libzkp/impl/cargo.lock/scroll-prover and halo2, contacted by a "-"
// The default `000000-000000` is set for integration test, and will be overwritten by coordinator's & prover's actual compilations (see their Makefiles).
var ZkVersion = "000000-000000"
```

or could we use the `tag` version?

**Comment by georgehao** [comment]
If we don't check VK version, is there have issue? only check the prover version equality. @HAOYUatHZ

**Comment by 0xmountaintop** [comment]
> > > > CheckScrollProverVersion
> > > 
> > > 
> > > Do you mean moving this check outside `if getTaskParameter.VK != b.vk {`?
> > > ```
> > > 	// if the prover has a different vk
> > > 	if getTaskParameter.VK != b.vk {
> > > 		// if the prover reports a different prover version
> > > 		if !version.CheckScrollProverVersion(proverVersion.(string)) {
> > > 			return nil, fmt.Errorf("incompatible prover version. please upgrade your prover, expect version: %s, actual version: %s", version.Version, proverVersion.(string))
> > > 		}
> > > 		// if the prover reports a same prover version
> > > 		return nil, fmt.Errorf("incompatible vk. please check your params files or config files")
> > > 	}
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Yeah, we can do that. And also, I remembered some similar discussions before during `sepolia testnet` testing, this will make only the latest version available. also cc @georgehao @Thegaram
> > 
> > 
> > I mean move this out and modify. so that we check version > XXX (this can solve "only the latest version available").
> 
> A question when implementing this method, the version compared in `CheckScrollProverVersion` is the prover version, does it have the feature of `monotonically increasing` based on some comparison rules?
> 
> code (the former part):
> 
> ```
> // ZkVersion is commit-id of common/libzkp/impl/cargo.lock/scroll-prover and halo2, contacted by a "-"
> // The default `000000-000000` is set for integration test, and will be overwritten by coordinator's & prover's actual compilations (see their Makefiles).
> var ZkVersion = "000000-000000"
> ```
> 
> or could we use the `tag` version?

 the zkversion is not increasing
Yes for your new check, we only check tag version not zkversion

**Comment by 0xmountaintop** [comment]
> If we don't check VK version, is there have issue? only check the prover version equality. @HAOYUatHZ

We'd better check both

**Comment by colinlyguo** [comment]
> > If we don't check VK version, is there have issue? only check the prover version equality. @HAOYUatHZ
> 
> We'd better check both

added in [77b92e1](https://github.com/scroll-tech/scroll/pull/1161/commits/77b92e16f61279949d43096c2f8e2d81dc796d5d)

File: common/version/prover_version.go [hunk 1]
```diff
@@ -19,3 +22,54 @@ func CheckScrollProverVersion(proverVersion string) bool {
...
 	return remote[2] == local[2]
 }
+
+// parseVersion takes a version string and returns its major, minor, and patch numbers.
+func parseVersion(version string) (major, minor, patch int) {
+	trimVersion := strings.TrimPrefix(version, "v")
+	if trimVersion == version {
+		log.Error("version does not start with v", "vesion", version)
+		return 0, 0, 0
+	}
```
Comment (georgehao) [inline common/version/prover_version.go:32]:
```suggestion
	if !strings.HasPrefix(version, 'v') {
		log.Error("version does not start with v", "vesion", version)
		return 0, 0, 0
	}
```

File: common/version/prover_version.go [hunk 2]
```diff
@@ -19,3 +22,54 @@ func CheckScrollProverVersion(proverVersion string) bool {
...
+	if err != nil {
+		log.Error("invalid patch version", "value", parts[2], "error", err)
+		return 0, 0, 0
+	}
+
+	return major, minor, patch
+}
+
+// CheckScrollRepoVersion checks if the proverVersion is at least the minimum required version.
+func CheckScrollRepoVersion(proverVersion, minVersion string) bool {
```
Comment (georgehao) [inline common/version/prover_version.go]: How about use `https://github.com/Masterminds/semver`

File: coordinator/internal/logic/provertask/prover_task.go [hunk 3]
```diff
@@ -68,13 +73,21 @@ func (b *BaseProverTask) checkParameter(ctx *gin.Context, getTaskParameter *coor
 		return nil, fmt.Errorf("incompatible vk. please check your params files or config files")
 	}
 
+	isBlocked, err := b.proverBlockListOrm.IsPublicKeyBlocked(ctx, publicKey.(string))
+	if err != nil {
+		return nil, fmt.Errorf("failed to check if the public key %s is blocked before assigning a chunk task, err: %w, proverName: %s, proverVersion: %s", publicKey, err, proverName, proverVersion)
```
Comment (georgehao) [inline coordinator/internal/logic/provertask/prover_task.go]:
```suggestion
		return nil, fmt.Errorf("failed to check whether the public key %s is blocked before assigning a chunk task, err: %w, proverName: %s, proverVersion: %s", publicKey, err, proverName, proverVersion)
```

File: common/version/prover_version.go (hunk 2 above)
Comment (colinlyguo) [reply common/version/prover_version.go]: makes sense, changed in [4207440](https://github.com/scroll-tech/scroll/pull/1161/commits/420744081dda148400d3d90581a497f0c41391e6)

**Review by georgehao — APPROVED** [review]

**Review by georgehao — APPROVED** [review]

**Review by 0xmountaintop — APPROVED** [review]
Though I think after having `CheckScrollRepoVersion` we don't need to block provers by blocklist any more,
but it's ok to keep this.

**Comment by georgehao** [comment]
> k provers by blocklist any more,

I think we should have a mechanism to actively restrict some prover to get task

**Pull request merged**